lib/ZcapClient.js

```diff
@@ -0,0 +1,228 @@
+   *
+   * @returns {ZcapClient} - The new ZcapClient instance.
+   */
+  constructor({
+    baseUrl, defaultHeaders = {}, agent, invocationSigner, delegationSigner
+  } = {}) {
+    this.baseUrl = baseUrl;
+    this.agent = agent;
+    this.defaultHeaders = {...DEFAULT_HEADERS, ...defaultHeaders};
+    this.invocationSigner = invocationSigner;
+    this.delegationSigner = delegationSigner;
+  }
+
+  /**
+   * Delegates an Authorization Capability to a specified delegatee.
+   *
+   * @param {object} options - The options to use.
+   * @param {string} [options.url] - The relative URL to invoke the
+   *   Authorization Capability against. Either `url` or `capability` must
+   *   be specified.
+   * @param {string} [options.capability] - The relative URL to invoke the
+   *   Authorization Capability against. Either `capability` or `url` must
+   *   be specified.
+   * @param {string} options.delegate - The URL identifying the entity to
+   *   delegate to.
+   * @param {string} [options.expires] - Optional expiration value for the
+   *   delegation. Default is 5 minutes into the future.
+   * @param {string|Array} [options.allowedActions] - Optional list of allowed
+   *   actions or string specifying allowed delegated action. Default: [] -
+   *   delegate all actions.
+   *
+   * @returns {object} - A promise that resolves to an HTTP response.
+   */
+  async delegate({
+    url, capability, delegate, expires, allowedActions = []
+  } = {}) {
+    let delegatedCapability;
+
+    // convert string value for allowedActions to array
+    allowedActions = (typeof allowedActions === 'string') ?
+      allowedActions = [allowedActions] : allowedActions;
+
+    // default expiration is 5 minutes in the future
+    const defaultExpires = new Date(Date.now() + 5 * 60 * 1000).toISOString();
+
+    if(url) {
+      // generate the root capability
+      delegatedCapability = {
+        '@context': SECURITY_CONTEXT_V2_URL,
+        id: `urn:zcap:${await generateId()}`,
+        parentCapability: 'urn:zcap:' + url,
+        invocationTarget: url,
+        controller: delegate,
+        expires: expires || defaultExpires
+      };
+    } else {
+      delegatedCapability = {
+        // use a provided capability
+        '@context': SECURITY_CONTEXT_V2_URL,
+        id: `urn:zcap:${await generateId()}`,
+        parentCapability: capability.id,
+        invocationTarget: capability.invocationTarget,
+        controller: delegate,
+        expires: expires || defaultExpires
+      };
+    }
+
+    if(allowedActions.length > 0) {
+      delegatedCapability.allowedActions = allowedActions;
+    }
+
+    const signedDelegatedCapability = await jsigs.sign(
+      delegatedCapability, {
+        suite: new Ed25519Signature2018({
+          signer: this.delegationSigner,
+          verificationMethod: this.delegationSigner.id
+        }),
+        purpose: new CapabilityDelegation({
+          capabilityChain: [
+            'urn:zcap:' + url
```

Comment:
```suggestion
            `urn:zcap:${url}`
```

Reply:
This isn't correct when `capability` is given. It only works for the root capability case. So it needs to be fixed (and needs a test).